appengine/src/main/java/com/google/googleinterns/gscribe/services/impl/TokenServiceImpl.java

```diff
@@ -16,30 +16,151 @@
+        return idToken.getPayload().getSubject();
+    }
+
+    /**
+     * Called to verify IDToken received from firebase functions
+     * As this function was originally intended for firebase function client so client id is added for that service
+     * This function takes as input the IDToken passed in header for authentication
+     * If the JWT clears all authentication checks then unique userID is extracted from the JWT and returned
+     *
+     * @param IDTokenString ( a JWT, web token signed by google )
+     * @return userID ( unique user ID for the user included in JWT )
+     * @throws GeneralSecurityException,IOException ( by google verifier, or reading credentials file errors )
+     * @throws InvalidRequestException              ( if the verification fails then returned token is null, If we receive null token then return this exception )
+     */
+    @Override
+    public String firebaseVerifyIDToken(String IDTokenString) throws GeneralSecurityException, IOException, InvalidRequestException {
+        final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+        String clientID = "201502787341-rqsisrvv0givo5agv86p44e2hjui05or.apps.googleusercontent.com";
 
-    @Inject
-    public TokenServiceImpl(TokenGenerationService tokenGenerationService, TokenVerificationService tokenVerificationService) {
-        this.tokenGenerationService = tokenGenerationService;
-        this.tokenVerificationService = tokenVerificationService;
+        final NetHttpTransport HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
+        GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(HTTP_TRANSPORT, JSON_FACTORY).setAudience(Collections.singletonList(clientID)).build();
+        GoogleIdToken idToken = verifier.verify(IDTokenString);
+        if (idToken == null) throw new InvalidRequestException("Authentication failed");
+        return idToken.getPayload().getSubject();
     }
 
+    /**
+     * Called when user is being authorized so new tokens are generated from authorization code
+     * Generates new accessToken and refreshToken from authorization code
+     * Needs credentials file to generate the tokens
+     *
+     * @param authCode ( authentication code )
+     * @return userToken object containing access token, refresh token and unique user Id
+     * @throws GeneralSecurityException,IOException ( thrown by NetHttpTransport, GoogleClientSecrets, GoogleTokenResponse or by invalid credentials file  )
+     * @throws InvalidRequestException              ( if the authorization code is invalid )
+     */
     @Override
-    public String verifyIDToken(String IDToken) {
-        return null;
+    public TokenResponse generateToken(String authCode) throws GeneralSecurityException, IOException, InvalidRequestException {
+        TokenResponse tokenResponse;
+        try {
+            final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
```

Comment:
Not a good idea to add everything in try catch, only at the line where exception can be thrown.

Reply:
Changed to add only function throwing exception into the try/catch.